lib/streamlit/external/langchain/streamlit_callback_handler.py

```diff
@@ -0,0 +1,434 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+LangChain CallbackHandler that prints to streamlit.
+
+This is a special API that's imported and used by LangChain itself. Any updates
+to the public API (the StreamlitCallbackHandler constructor, and the entirety
+of LLMThoughtLabeler) *must* remain backwards-compatible to avoid breaking
+LangChain.
+
+This means that it's acceptable to add new optional kwargs to StreamlitCallbackHandler,
+but no new positional args or required kwargs should be added, and no existing
+args should be removed. If we need to overhaul the API, we must ensure that a
+compatible API continues to exist.
+
+Any major change to the StreamlitCallbackHandler should be tested by importing
+the API *from LangChain itself*.
+"""
+
+from __future__ import annotations
+
+from enum import Enum
+from typing import Any, Dict, List, NamedTuple, Optional, Union
+
+from langchain.callbacks.base import BaseCallbackHandler
+from langchain.schema import AgentAction, AgentFinish, LLMResult
+
+from streamlit.delta_generator import DeltaGenerator
+from streamlit.external.langchain.mutable_expander import MutableExpander
+
+
+def _convert_newlines(text: str) -> str:
+    """Convert newline characters to markdown newline sequences
+    (space, space, newline).
+    """
+    return text.replace("\n", "  \n")
+
+
+CHECKMARK_EMOJI = "✅"
+THINKING_EMOJI = ":thinking_face:"
+HISTORY_EMOJI = ":books:"
+EXCEPTION_EMOJI = "⚠️"
+
+
+class LLMThoughtState(Enum):
+    # The LLM is thinking about what to do next. We don't know which tool we'll run.
+    THINKING = "THINKING"
+    # The LLM has decided to run a tool. We don't have results from the tool yet.
+    RUNNING_TOOL = "RUNNING_TOOL"
+    # We have results from the tool.
+    COMPLETE = "COMPLETE"
+
+
+class ToolRecord(NamedTuple):
+    name: str
+    input_str: str
+
+
+class LLMThoughtLabeler:
```

Comment:
Yeah - the primary use case was for the frosty demo to provide nicer display names for its tools (so that you'd see e.g. "Searching" instead of "search_tool"). You could use translate into other languages as well.